### PR TITLE
[Arm64] ASIMD Shift instructions

### LINKED ISSUE #35702: Stabilize genArm64EmitterUnitTests
While working recently on adding unit test inside `genArm64EmitterUnitTests()` I found out that we hit various asserts in the unit test, probably because we changed the implementation of emitter methods but didn't go back and update the test. Lets take some time to clean it up.

### PR DESCRIPTION
**Pull request by echesakov**
1) Adds ASIMD shift instructions that are needed for #33398:
* sqrshl
* sqrshrn
* sqrshrn2
* sqrshrun
* sqrshrun2
* sqshl
* sqshlu
* sqshrn
* sqshrn2
* sqshrun
* sqshrun2
* srshl
* sshl
* uqrshl
* uqrshrn
* uqrshrn2
* uqshl
* uqshrn
* uqshrn2
* urshl
* ushl

2) Turns out that our implementation of insEncodeVectorShift worked only for SIMD left shift (immediate) operations while for SIMD right shift (immediate) - there is a different formula for computing a corresponding immediate value (since for right shifts the immediate range is `[1, DataSizeInBits]`, while for left shifts this is `[0, DataSizeInBits)`). I extended insEncodeVectorShift to support right shifts

3) Fixes tests failures in #35702:
* tbx and tbl (51650376fb5c768f06781822edcb380f2383f954) 
* fcvtl{2} and fcvtn{2} - those were implemented as "scalar" while, in fact, they are "vector" operations. Fixed this in 28690e95f888ab59ad4eabcfebdcf7364a54a19b and update tests with d6c6c4bdadc65f53235c0a790750166800f5d1b2
* Also making them real "vector" instructions in 28690e95f888ab59ad4eabcfebdcf7364a54a19b allows to simplify genSIMDIntrinsicWiden (5db367f1141ef451af1010a7f144183c21e6ba2f) and genSIMDIntrinsicNarrow(d2b20e61ee868225960520b6cd31845457c1184e) since now there instructions follow the same protocol as other Widenning or Narrowing instructions (i.e. expect `opt` to be non-NONE, `size` to correspond to the `opt`)

4) Lots of *cosmetic* changes in instrsarm64.h - since some of the new instruction have quite long opcode names - I re-formatted the file to make it look less messy. In addition, I removed fp which is never used on Arm64 and always set to 0.

5) Removed references to Arm Architecuture Reference Manual - I updated the document and found that they added new instructions and changed the paragraphs numbering.

Fixes #35702 

Attached are jitDisasm and WinDbg u command outputs.
[jitDisasm](https://github.com/dotnet/runtime/files/4637183/jitDisasm.txt)
[WinDbg u](https://github.com/dotnet/runtime/files/4637184/windbg-u.txt)

### CODE REVIEW

**Comment by echesakovMSFT** [comment]
@BruceForstall @CarolEidt PTAL

cc @dotnet/jit-contrib

**Comment by BruceForstall** [comment]
I noticed some interesting differences between the JIT disasm and windbg disassembly (not all related to this change):

```
JIT:
eb29011f cmp x8,x9,uxtb
eb29211f cmp x8,x9,uxth
eb29811f cmp x8,x9,sxtb
eb29a11f cmp x8,x9,sxth
eb29051f cmp x8,x9,uxtb #1
eb29291f cmp x8,x9,uxth #2
eb29851f cmp x8,x9,sxtb #1
eb29a91f cmp x8,x9,sxth #2
6b29011f cmp w8,x9,uxtb
6b29291f cmp w8,x9,uxth #2
6b29811f cmp w8,x9,sxtb
6b29a91f cmp w8,x9,sxth #2
8b2a0128 add x8,x9,x10,uxtb
8b2a2128 add x8,x9,x10,uxth
8b2a8128 add x8,x9,x10,sxtb
8b2aa128 add x8,x9,x10,sxth
8b2a1128 add x8,x9,x10,uxtb #4
8b2a3128 add x8,x9,x10,uxth #4
8b2a9128 add x8,x9,x10,sxtb #4
8b2ab128 add x8,x9,x10,sxth #4
4b2a0128 sub w8,w9,x10,uxtb
4b2a2128 sub w8,w9,x10,uxth
4b2a6128 sub w8,w9,x10,uxtx
4b2a8128 sub w8,w9,x10,sxtb
4b2aa128 sub w8,w9,x10,sxth
4b2ae128 sub w8,w9,x10,sxtx
4b2a1128 sub w8,w9,x10,uxtb #4
4b2a3128 sub w8,w9,x10,uxth #4
4b2a7128 sub w8,w9,x10,uxtx #4
4b2a9128 sub w8,w9,x10,sxtb #4
4b2ab128 sub w8,w9,x10,sxth #4
4b2af128 sub w8,w9,x10,sxtx #4

windbg:
eb29011f cmp x8,w9,uxtb
eb29211f cmp x8,w9,uxth
eb29811f cmp x8,w9,sxtb
eb29a11f cmp x8,w9,sxth
eb29051f cmp x8,w9,uxtb #1
eb29291f cmp x8,w9,uxth #2
eb29851f cmp x8,w9,sxtb #1
eb29a91f cmp x8,w9,sxth #2
6b29011f cmp w8,w9,uxtb
6b29291f cmp w8,w9,uxth #2
6b29811f cmp w8,w9,sxtb
6b29a91f cmp w8,w9,sxth #2
8b2a0128 add x8,x9,w10,uxtb
8b2a2128 add x8,x9,w10,uxth
8b2a8128 add x8,x9,w10,sxtb
8b2aa128 add x8,x9,w10,sxth
8b2a1128 add x8,x9,w10,uxtb #4
8b2a3128 add x8,x9,w10,uxth #4
8b2a9128 add x8,x9,w10,sxtb #4
8b2ab128 add x8,x9,w10,sxth #4
4b2a0128 sub w8,w9,w10,uxtb
4b2a2128 sub w8,w9,w10,uxth
4b2a6128 sub w8,w9,w10,uxtx
4b2a8128 sub w8,w9,w10,sxtb
4b2aa128 sub w8,w9,w10,sxth
4b2ae128 sub w8,w9,w10,sxtx
4b2a1128 sub w8,w9,w10,uxtb #4
4b2a3128 sub w8,w9,w10,uxth #4
4b2a7128 sub w8,w9,w10,uxtx #4
4b2a9128 sub w8,w9,w10,sxtb #4
4b2ab128 sub w8,w9,w10,sxth #4
4b2af128 sub w8,w9,w10,sxtx #4
```
in the `uxtw` case, the JIT emits w9, but in the `uxtb`/`uxth` cases, it incorrectly emits `x9` instead of `w9`.

Similar:
```
JIT:
9b2afd28 smnegl x8,x9,x10
9baa7d28 umull x8,x9,x10
9baafd28 umnegl x8,x9,x10
9b2afd28 smnegl w8,w9,w10
9baa7d28 umull w8,w9,w10
9baafd28 umnegl w8,w9,w10
9bca7d28 umulh w8,w9,w10
9b2031c2 smaddl w2,w14,w0,w12
9b21b5e3 smsubl w3,w15,w1,w13
9ba23a64 umaddl w4,w19,w2,w14
9ba3be85 umsubl w5,w20,w3,w15
9b2656e8 smaddl x8,x23,x6,x21
9b27db09 smsubl x9,x24,x7,x22
9ba85f2a umaddl x10,x25,x8,x23
9ba9e34b umsubl x11,x26,x9,x24

windbg:
9b2afd28 smnegl x8,w9,w10
9baa7d28 umull x8,w9,w10
9baafd28 umnegl x8,w9,w10
9b2afd28 smnegl x8,w9,w10
9baa7d28 umull x8,w9,w10
9baafd28 umnegl x8,w9,w10
9bca7d28 umulh x8,x9,x10
9b2031c2 smaddl x2,w14,w0,x12
9b21b5e3 smsubl x3,w15,w1,x13
9ba23a64 umaddl x4,w19,w2,x14
9ba3be85 umsubl x5,w20,w3,x15
9b2656e8 smaddl x8,w23,w6,x21
9b27db09 smsubl x9,w24,w7,x22
9ba85f2a umaddl x10,w25,w8,x23
9ba9e34b umsubl x11,w26,w9,x24
```

Your windbg is missing decoding for many instructions:
```
08a87d49 casb w8,w9,[x10]
08e87d49 casab w8,w9,[x10]
08e8fd49 casalb w8,w9,[x10]
08a8fd49 caslb w8,w9,[x10]
48a87d49 cash w8,w9,[x10]
48e87d49 casah w8,w9,[x10]
48e8fd49 casalh w8,w9,[x10]
48a8fd49 caslh w8,w9,[x10]
88a87d49 cas w8,w9,[x10]
88e87d49 casa w8,w9,[x10]
88e8fd49 casal w8,w9,[x10]
88a8fd49 casl w8,w9,[x10]
c8a87d49 cas x8,x9,[x10]
c8e87d49 casa x8,x9,[x10]
c8e8fd49 casal x8,x9,[x10]
c8a8fd49 casl x8,x9,[x10]
38280149 ldaddb w8,w9,[x10]
38a80149 ldaddab w8,w9,[x10]
38e80149 ldaddalb w8,w9,[x10]
38680149 ldaddlb w8,w9,[x10]
78280149 ldaddh w8,w9,[x10]
78a80149 ldaddah w8,w9,[x10]
78e80149 ldaddalh w8,w9,[x10]
78680149 ldaddlh w8,w9,[x10]
b8280149 ldadd w8,w9,[x10]
b8a80149 ldadda w8,w9,[x10]
b8e80149 ldaddal w8,w9,[x10]
b8680149 ldaddl w8,w9,[x10]
f8280149 ldadd x8,x9,[x10]
f8a80149 ldadda x8,x9,[x10]
f8e80149 ldaddal x8,x9,[x10]
f8680149 ldaddl x8,x9,[x10]
38288149 swpb w8,w9,[x10]
38a88149 swpab w8,w9,[x10]
38e88149 swpalb w8,w9,[x10]
38688149 swplb w8,w9,[x10]
78288149 swph w8,w9,[x10]
78a88149 swpah w8,w9,[x10]
78e88149 swpalh w8,w9,[x10]
78688149 swplh w8,w9,[x10]
b8288149 swp w8,w9,[x10]
b8a88149 swpa w8,w9,[x10]
b8e88149 swpal w8,w9,[x10]
b8688149 swpl w8,w9,[x10]
f8288149 swp x8,x9,[x10]
f8a88149 swpa x8,x9,[x10]
f8e88149 swpal x8,x9,[x10]
f8688149 swpl x8,x9,[x10]
3828015f ldaddb w8,wzr,[x10]
3868015f ldaddlb w8,wzr,[x10]
7828015f ldaddh w8,wzr,[x10]
7868015f ldaddlh w8,wzr,[x10]
b828015f ldadd w8,wzr,[x10]
b868015f ldaddl w8,wzr,[x10]
f828015f ldadd x8,xzr,[x10]
f868015f ldaddl x8,xzr,[x10]
```
Are you up to date? Does it not support ARMv8.1 and beyond?

These have the wrong destination width:
```
JIT:
0e3038a4 saddlv b4,v5.8b
4e3038e6 saddlv b6,v7.16b
0e703928 saddlv h8,v9.4h
4e70396a saddlv h10,v11.8h
4eb039ee saddlv s14,v15.4s
2e3038a4 uaddlv b4,v5.8b
6e3038e6 uaddlv b6,v7.16b
2e703928 uaddlv h8,v9.4h
6e70396a uaddlv h10,v11.8h
6eb039ee uaddlv s14,v15.4s

windbg:
0e3038a4 saddlv h4,v5.8b
4e3038e6 saddlv h6,v7.16b
0e703928 saddlv s8,v9.4h
4e70396a saddlv s10,v11.8h
4eb039ee saddlv d14,v15.4s
2e3038a4 uaddlv h4,v5.8b
6e3038e6 uaddlv h6,v7.16b
2e703928 uaddlv s8,v9.4h
6e70396a uaddlv s10,v11.8h
6eb039ee uaddlv d14,v15.4s
```

These have an incorrect source width:
```
JIT:
5f0f9c20 sqrshrn b0,b1,#1
5f089c62 sqrshrn b2,b3,#8
5f179ca4 sqrshrn h4,h5,#9
5f109ce6 sqrshrn h6,h7,#16
5f2f9d28 sqrshrn s8,s9,#17
5f209d6a sqrshrn s10,s11,#32
7f0f8c20 sqrshrun b0,b1,#1
7f088c20 sqrshrun b0,b1,#8
7f178c62 sqrshrun h2,h3,#9
7f108c62 sqrshrun h2,h3,#16
7f2f8ca4 sqrshrun s4,s5,#17
7f208ca4 sqrshrun s4,s5,#32
5f0f9420 sqshrn b0,b1,#1
5f089462 sqshrn b2,b3,#8
5f1794a4 sqshrn h4,h5,#9
5f1094e6 sqshrn h6,h7,#16
5f2f9528 sqshrn s8,s9,#17
5f20956a sqshrn s10,s11,#32
7f0f8420 sqshrun b0,b1,#1
7f088462 sqshrun b2,b3,#8
7f1784a4 sqshrun h4,h5,#9
7f1084e6 sqshrun h6,h7,#16
7f2f8528 sqshrun s8,s9,#17
7f20856a sqshrun s10,s11,#32
7f0f9c20 uqrshrn b0,b1,#1
7f089c62 uqrshrn b2,b3,#8
7f179ca4 uqrshrn h4,h5,#9
7f109ce6 uqrshrn h6,h7,#16
7f2f9d28 uqrshrn s8,s9,#17
7f209d6a uqrshrn s10,s11,#32
7f0f9420 uqshrn b0,b1,#1
7f089462 uqshrn b2,b3,#8
7f1794a4 uqshrn h4,h5,#9
7f1094e6 uqshrn h6,h7,#16
7f2f9528 uqshrn s8,s9,#17
7f20956a uqshrn s10,s11,#32

windbg:
5f0f9c20 sqrshrn b0,h1,#1
5f089c62 sqrshrn b2,h3,#8
5f179ca4 sqrshrn h4,s5,#9
5f109ce6 sqrshrn h6,s7,#16
5f2f9d28 sqrshrn s8,d9,#17
5f209d6a sqrshrn s10,d11,#32
7f0f8c20 sqrshrun b0,h1,#1
7f088c20 sqrshrun b0,h1,#8
7f178c62 sqrshrun h2,s3,#9
7f108c62 sqrshrun h2,s3,#16
7f2f8ca4 sqrshrun s4,d5,#17
7f208ca4 sqrshrun s4,d5,#32
5f0f9420 sqshrn b0,h1,#1
5f089462 sqshrn b2,h3,#8
5f1794a4 sqshrn h4,s5,#9
5f1094e6 sqshrn h6,s7,#16
5f2f9528 sqshrn s8,d9,#17
5f20956a sqshrn s10,d11,#32
7f0f8420 sqshrun b0,h1,#1
7f088462 sqshrun b2,h3,#8
7f1784a4 sqshrun h4,s5,#9
7f1084e6 sqshrun h6,s7,#16
7f2f8528 sqshrun s8,d9,#17
7f20856a sqshrun s10,d11,#32
7f0f9c20 uqrshrn b0,h1,#1
7f089c62 uqrshrn b2,h3,#8
7f179ca4 uqrshrn h4,s5,#9
7f109ce6 uqrshrn h6,s7,#16
7f2f9d28 uqrshrn s8,d9,#17
7f209d6a uqrshrn s10,d11,#32
7f0f9420 uqshrn b0,h1,#1
7f089462 uqshrn b2,h3,#8
7f1794a4 uqshrn h4,s5,#9
7f1094e6 uqshrn h6,s7,#16
7f2f9528 uqshrn s8,d9,#17
7f20956a uqshrn s10,d11,#32
```

File: src/coreclr/src/jit/codegenarm64.cpp
Comment (BruceForstall) [inline src/coreclr/src/jit/codegenarm64.cpp:9080]: You need a 2D case here

File: src/coreclr/src/jit/codegenarm64.cpp
Comment (BruceForstall) [inline src/coreclr/src/jit/codegenarm64.cpp:9094]: +2D

File: src/coreclr/src/jit/codegenarm64.cpp
Comment (BruceForstall) [inline src/coreclr/src/jit/codegenarm64.cpp:8418]: It would be useful to comment here that you have two of each type, one with small constant and one with large, because the constant is split in the encoding, and you want to verify that the split constant encoding works.

**Review by BruceForstall — CHANGES_REQUESTED** [review]
A few comments.

**Comment by echesakovMSFT** [comment]
@BruceForstall Thank you for thorough review! I believe I have addressed all your feedback except one - regarding incorrect source width for sqshrn,  sqrshrun, uqshrn. This one is slightly harder to fix and need some time to figure out how. 

However, I would like to unblock my ASIMD Shift intrinsics work. Would it be ok for you if I open a separate GitHub issue and follow up later on? 

BTW, I have updated the WinDbg and now all the atomic operations are disassembled.
I attached the output files.

[windbg-u.txt](https://github.com/dotnet/runtime/files/4660391/windbg-u.txt)
[jitDisasm.txt](https://github.com/dotnet/runtime/files/4660392/jitDisasm.txt)

File: src/coreclr/src/jit/emitarm64.cpp
Comment (CarolEidt) [inline src/coreclr/src/jit/emitarm64.cpp:8860]: Minor: This header is missing the `//----` line.

**Review by CarolEidt — APPROVED** [review]
LGTM - only one optional formatting thing (wouldn't hold up this PR for it).
There's a *lot* of detail here, so I'm not sure I scrutinized it all, but I really appreciate all the cleanup!

**Comment by BruceForstall** [comment]
I'm ok with opening an issue for the remaining disassembly issues. From your attachments, it looks like there are still inconsistencies/errors with `cmp`/`add`/`sub` with extensions, `smnegl`, `umull`, `umnegl`, `umulh`, `smaddl`, `smsubl`, `umaddl`, `umsubl`.

**Review by BruceForstall — APPROVED** [review]

**Comment by echesakovMSFT** [comment]
@BruceForstall I will take a look at these inconsistencies

**Pull request merged**
